packages/eslint-plugin/src/rules/no-enum-literals.ts

```diff
@@ -0,0 +1,104 @@
+import ts from 'typescript';
+import * as util from '../util';
+import { TSESTree } from '@typescript-eslint/typescript-estree';
+
+export default util.createRule({
+  name: 'no-enum-literals',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallows usage of literals instead of enums',
+      category: 'Best Practices',
+      recommended: 'error',
+    },
+    messages: {
+      noLiterals: 'Do not use literal values instead of enums',
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false,
+      },
+    ],
+  },
+  defaultOptions: [{}],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    /**
+     * is node an identifier with type of an enum
+     * @param node identifier node
+     */
+    function isNodeEnumIdentifier(node: TSESTree.Node): boolean {
+      const originalNode = parserServices.esTreeNodeToTSNodeMap.get<
+        ts.Identifier
+      >(node);
+      const type = checker.getTypeAtLocation(originalNode);
+
+      if (!type.symbol) {
+        return false;
+      }
+
+      const { name } = type.symbol;
+
+      return !['Number', 'String'].includes(name);
```

Comment:
I was going to say that this will be a lot slower than just writing the two boolean checks by hand, but running a jsperf says that it's actually the same speed.
I suspect that the js engine unrolls this into a set of boolean checks behind the scenes.
#TodayILearned

https://jsperf.com/array-includes-vs-chained-booleans/1

Reply:
However, `includes` is a bit slower when it returns `false`: https://jsperf.com/array-includes-vs-chained-booleans/3